File: sync/src/kite_jumps.py
"""Extract kiteboarding jump data from Garmin/Surfr FIT files.

The watch runs Surfr's Connect IQ data field, which writes per-jump height into
the FIT `record` stream (developer field `heights`) together with GPS position
and timestamp. We validated that this per-record height matches Surfr's
cloud-corrected value exactly (e.g. 5.20/4.74/4.23/3.78/3.69 m on Schinias
2025-11-18). The session-level summary `maxheight` field is unreliable (it can
glitch to e.g. 13.51 m) and is intentionally ignored.

Surfr's auto-exported Strava activity (stored in `strava_raw_data`) carries the
corrected Top-5 description with per-jump airtime / distance / approach-speed.
We join that in by rank to enrich the biggest jumps.

Heights stay in meters (the kiteboarding convention for jump height). Speeds are
exposed in knots per the project preference.
"""

import re

import fitparse

KMH_TO_KNOTS = 1 / 1.852
MS_TO_KNOTS = 3.6 / 1.852

# A record's `heights` field is non-zero only during a jump; the same height
# repeats across the airtime samples. Runs separated by more than this many
# seconds are treated as separate jumps.
_JUMP_GAP_S = 4
_MIN_JUMP_M = 0.3
# Surfr only corrects heights it agrees with within this tolerance; used to
# guard the rank-based join so a mismatched session can't attach wrong airtimes.
_SURFR_MATCH_TOL_M = 0.5

_SURFR_JUMP_RE = re.compile(
    r"Nr\.\s*(\d+)\s*,\s*H:\s*([\d.]+)\s*m\s*,\s*A:\s*([\d.]+)\s*sec\.?\s*,\s*"
    r"D:\s*(\d+)\s*m\s*,\s*Max\s*Speed:\s*(\d+)\s*Km/h",
    re.IGNORECASE,
)
_SURFR_SPOT_RE = re.compile(r"Spot:\s*'([^']+)'", re.IGNORECASE)
_SURFR_MAXAIR_RE = re.compile(r"Max\.?\s*Airtime:\s*([\d.]+)\s*sec", re.IGNORECASE)


def _semicircles_to_deg(v):
    return v * (180 / 2**31) if v is not None else None


def _read_jump_samples(fit_path: str) -> list[dict]:
    """Read the per-record jump samples (heights>0) from a FIT file.

    Besides the `heights` value, each record carries three richer developer
    fields written by Surfr's Connect IQ app (validated against Surfr's
    cloud-corrected output — exact match):
      - `jump`:      (height_m, airtime_s, distance_m, approach_speed_ms)
      - `jumpchart`: uint32[20] — the measured height trajectory in cm
      - `timestamps`: (detect_ts, takeoff_ts, landing_ts) unix seconds
    """
    fit = fitparse.FitFile(fit_path)
    samples = []
    for msg in fit.get_messages("record"):
        d = {x.name: x.value for x in msg}
        h = d.get("heights")
        if isinstance(h, (int, float)) and h and h > _MIN_JUMP_M:
            samples.append(
                {
                    "time": d.get("timestamp"),
                    "height_m": float(h),
                    "lat": _semicircles_to_deg(d.get("position_lat")),
                    "lng": _semicircles_to_deg(d.get("position_long")),
                    "speed_ms": d.get("enhanced_speed"),
                    "jump_tuple": d.get("jump"),
                    "jumpchart": d.get("jumpchart"),
                    "jump_ts": d.get("timestamps"),
                }
            )
    return samples


def group_jumps(samples: list[dict]) -> list[dict]:
    """Group time-ordered jump samples into individual jumps (peak per run),
    sorted biggest-first. Pure function over sample dicts."""
    groups: list[list[dict]] = []
    cur: list[dict] = []
    for s in samples:
        if not cur or (s["time"] - cur[-1]["time"]).total_seconds() <= _JUMP_GAP_S:
            cur.append(s)
        else:
            groups.append(cur)
            cur = [s]
    if cur:
        groups.append(cur)

    jumps = []
    for g in groups:
        peak = max(g, key=lambda x: x["height_m"])
        jump = {
            "height_m": round(peak["height_m"], 2),
            "lat": round(peak["lat"], 6) if peak["lat"] is not None else None,
            "lng": round(peak["lng"], 6) if peak["lng"] is not None else None,
            "time": peak["time"].isoformat() if hasattr(peak["time"], "isoformat") else peak["time"],
        }
        # Rich per-jump data from the Surfr CIQ developer fields (exact match with
        # Surfr's corrected output — validated). Falls back gracefully when absent.
        tup = peak.get("jump_tuple")
        if tup and len(tup) >= 4 and tup[0] and tup[0] > 0:
            jump["airtime_s"] = round(float(tup[1]), 2)
            jump["distance_m"] = round(float(tup[2]), 1)
            jump["approach_speed_kn"] = round(float(tup[3]) * MS_TO_KNOTS, 1)
        elif peak.get("speed_ms"):
            # NB: never derive airtime from the heights-run span — it lingers past
            # the real airtime and overestimates badly.
            jump["approach_speed_kn"] = round(peak["speed_ms"] * MS_TO_KNOTS, 1)
        chart = peak.get("jumpchart")
        if chart:
            vals = [v / 100 for v in chart]
            last = max((i for i, v in enumerate(vals) if v > 0), default=-1)
            if last >= 0:
                # Keep one bounding zero on each side (takeoff/landing on the water).
                jump["trajectory_m"] = [0.0] + [round(v, 2) for v in vals[1 : last + 1]] + [0.0]
        ts = peak.get("jump_ts")
        if ts and len(ts) >= 3 and ts[1] and ts[2]:
            jump["takeoff_ts"] = int(ts[1])
            jump["landing_ts"] = int(ts[2])
        jumps.append(jump)
    jumps.sort(key=lambda j: j["height_m"], reverse=True)
    return jumps


# Seconds of on-water context captured around each jump's flight path.
_PATH_LEAD_S = 3


def _read_track(fit_path: str) -> list[tuple[int, float, float]]:
    """(unix_ts, lat, lng) for every 1Hz record with a GPS fix.

    FIT record timestamps are NAIVE UTC datetimes; `.timestamp()` on a naive
    value assumes local time and shifts by the machine's UTC offset, breaking
    the match against the dev-field unix timestamps. Force UTC explicitly.
    """
    from datetime import timezone

    fit = fitparse.FitFile(fit_path)
    track = []
    for msg in fit.get_messages("record"):
        d = {x.name: x.value for x in msg}
        lat = _semicircles_to_deg(d.get("position_lat"))
        lng = _semicircles_to_deg(d.get("position_long"))
        t = d.get("timestamp")
        if lat and lng and t is not None:
            track.append((int(t.replace(tzinfo=timezone.utc).timestamp()), lat, lng))
    return track


def attach_jump_paths(jumps: list[dict], track: list[tuple[int, float, float]], top_n: int = 3) -> None:
    """Attach the real flight-window GPS path (takeoff-3s .. landing+3s, 1Hz) to
    the top N jumps: `path` = [[t_rel_s_from_takeoff, lat, lng], ...]."""
    if not track:
        return
    for j in jumps[:top_n]:
        t0, t1 = j.get("takeoff_ts"), j.get("landing_ts")
        if not t0 or not t1:
            continue
        window = [
            [ts - t0, round(lat, 6), round(lng, 6)]
            for ts, lat, lng in track
            if t0 - _PATH_LEAD_S <= ts <= t1 + _PATH_LEAD_S
        ]
        if len(window) >= 3:
            j["path"] = window


def extract_jumps_from_fit(fit_path: str) -> list[dict]:
    """Per-jump {height_m, lat, lng, time, airtime, trajectory, path} from a FIT,
    biggest-first.

    Uses only the per-record `heights` developer field, never the glitchy
    session-level `maxheight`.
    """
    jumps = group_jumps(_read_jump_samples(fit_path))
    attach_jump_paths(jumps, _read_track(fit_path))
    return jumps


def parse_surfr_description(description: str | None) -> list[dict]:
    """Parse Surfr's 'Top 5 Jumps' block into per-jump corrected detail."""
    if not description:
        return []
    out = []
    for m in _SURFR_JUMP_RE.finditer(description):
        rank, h, air, dist, kmh = m.groups()
        out.append(
            {
                "rank": int(rank),
                "height_m": float(h),
                "airtime_s": float(air),
                "distance_m": int(dist),
                "approach_speed_kn": round(int(kmh) * KMH_TO_KNOTS, 1),
            }
        )
    return out


def assemble_jumps(jumps: list[dict], surfr_description: str | None = None) -> dict:
    """Rank jumps, enrich the top ones with Surfr airtime/distance (matched by
    rank, height-guarded), and compute the session summary. Pure function."""
    surfr = parse_surfr_description(surfr_description)
    for i, j in enumerate(jumps):
        j["rank"] = i + 1
        s = next((s for s in surfr if s["rank"] == i + 1), None)
        if s and abs(s["height_m"] - j["height_m"]) <= _SURFR_MATCH_TOL_M:
            # Native airtime/distance (from the CIQ `jump` tuple) match Surfr's
            # corrected values exactly, so Surfr's text is only a fallback for
            # old sessions plus a cross-check marker.
            j.setdefault("airtime_s", s["airtime_s"])
            j.setdefault("distance_m", s["distance_m"])
            j["surfr_height_m"] = s["height_m"]

    # Spot name + session max-airtime from Surfr's description when available.
    spot = None
    surfr_max_air = None
    if surfr_description:
        m = _SURFR_SPOT_RE.search(surfr_description)
        if m:
            spot = m.group(1)
        m = _SURFR_MAXAIR_RE.search(surfr_description)
        if m:
            surfr_max_air = float(m.group(1))
    # Native per-jump airtimes are Surfr-accurate; Surfr's text is the fallback.
    airtimes = [j["airtime_s"] for j in jumps if j.get("airtime_s")]
    max_airtime = max(airtimes) if airtimes else surfr_max_air

    heights = [j["height_m"] for j in jumps]
    summary = {
        "jump_count": len(jumps),
        "max_height_m": max(heights) if heights else None,
        "avg_height_m": round(sum(heights) / len(heights), 2) if heights else None,
        "total_height_m": round(sum(heights), 1) if heights else None,
        "max_airtime_s": max_airtime,
        "spot": spot,
        "surfr_matched": bool(surfr),
    }
    return {"jumps": jumps, "summary": summary}


def build_kite_jumps(fit_path: str, surfr_description: str | None = None) -> dict:
    """Assemble the stored jump payload from a FIT path + optional Surfr text."""
    return assemble_jumps(extract_jumps_from_fit(fit_path), surfr_description)

File: web/lib/kite-jumps.ts
/**
 * Kiteboarding jump extraction from Garmin/Surfr FIT files — TS port of
 * sync/src/kite_jumps.py. Reads Surfr's Connect IQ per-record developer fields
 * (heights, jump, jumpchart, timestamps) via @garmin/fitsdk, groups them into
 * jumps (peak per run), attaches flight-path GPS windows, and enriches the top
 * jumps from Surfr's Strava description. Stage: sync cutover (#187).
 *
 * Heights stay in meters; speeds are exposed in knots (project convention).
 */
import { Decoder, Stream } from "@garmin/fitsdk";
import { unzipSync } from "fflate";
import type { GarminClient } from "garmin-auth";
import type { QueryFn } from "./db";

const KMH_TO_KNOTS = 1 / 1.852;
const MS_TO_KNOTS = 3.6 / 1.852;
const JUMP_GAP_S = 4;
const MIN_JUMP_M = 0.3;
const SURFR_MATCH_TOL_M = 0.5;
const PATH_LEAD_S = 3;

// Python round() parity: toFixed matches except exact binary ties.
const r = (x: number, n: number): number => {
  const m = 10 ** n;
  const twice = x * 2 * m;
  if (Number.isInteger(twice) && Math.abs(twice) % 2 === 1) {
    const floor = Math.floor(x * m);
    return (floor % 2 === 0 ? floor : floor + 1) / m;
  }
  return Number(x.toFixed(n));
};

const SURFR_JUMP_RE = /Nr\.\s*(\d+)\s*,\s*H:\s*([\d.]+)\s*m\s*,\s*A:\s*([\d.]+)\s*sec\.?\s*,\s*D:\s*(\d+)\s*m\s*,\s*Max\s*Speed:\s*(\d+)\s*Km\/h/gi;
const SURFR_SPOT_RE = /Spot:\s*'([^']+)'/i;
const SURFR_MAXAIR_RE = /Max\.?\s*Airtime:\s*([\d.]+)\s*sec/i;

function semicirclesToDeg(v: number | null | undefined): number | null {
  return v == null ? null : v * (180 / 2 ** 31);
}

interface JumpSample {
  time: Date;
  height_m: number;
  lat: number | null;
  lng: number | null;
  speed_ms: number | null;
  jump_tuple: number[] | null;
  jumpchart: number[] | null;
  jump_ts: number[] | null;
}

export interface Jump {
  height_m: number;
  lat: number | null;
  lng: number | null;
  time: string;
  airtime_s?: number;
  distance_m?: number;
  approach_speed_kn?: number;
  trajectory_m?: number[];
  takeoff_ts?: number;
  landing_ts?: number;
  rank?: number;
  surfr_height_m?: number;
  path?: Array<[number, number, number]>;
}

/** Decode a FIT file; return record messages + a developer-field name→index map. */
function decodeRecords(bytes: Uint8Array): { records: any[]; devIdx: Record<string, number> } {
  const decoder = new Decoder(Stream.fromByteArray(bytes));
  const { messages } = decoder.read({ convertDateTimesToDates: true });
  const fds = (messages.fieldDescriptionMesgs || []) as any[];
  // fitsdk keys record developerFields by the field_description definition order.
  const devIdx: Record<string, number> = {};
  fds.forEach((f, i) => { devIdx[f.fieldName] = i; });
  return { records: (messages.recordMesgs || []) as any[], devIdx };
}

/** Per-record jump samples (heights > MIN). Port of _read_jump_samples. */
function readJumpSamples(records: any[], devIdx: Record<string, number>): JumpSample[] {
  const iH = devIdx.heights, iJ = devIdx.jump, iC = devIdx.jumpchart, iT = devIdx.timestamps;
  const samples: JumpSample[] = [];
  for (const rec of records) {
    const dev = rec.developerFields || {};
    const h = dev[iH];
    if (typeof h === "number" && h && h > MIN_JUMP_M) {
      samples.push({
        time: rec.timestamp,
        height_m: Number(h),
        lat: semicirclesToDeg(rec.positionLat),
        lng: semicirclesToDeg(rec.positionLong),
        speed_ms: rec.enhancedSpeed ?? null,
        jump_tuple: (dev[iJ] as number[]) ?? null,
        jumpchart: (dev[iC] as number[]) ?? null,
        jump_ts: (dev[iT] as number[]) ?? null,
      });
    }
  }
  return samples;
}

/** Group time-ordered samples into jumps (peak per run), biggest-first. Pure. */
export function groupJumps(samples: JumpSample[]): Jump[] {
  const groups: JumpSample[][] = [];
  let cur: JumpSample[] = [];
  for (const s of samples) {
    if (!cur.length || (s.time.getTime() - cur[cur.length - 1].time.getTime()) / 1000 <= JUMP_GAP_S) {
      cur.push(s);
    } else {
      groups.push(cur);
      cur = [s];
    }
  }
  if (cur.length) groups.push(cur);

  const jumps: Jump[] = [];
  for (const g of groups) {
    const peak = g.reduce((a, b) => (b.height_m > a.height_m ? b : a));
    const jump: Jump = {
      height_m: r(peak.height_m, 2),
      lat: peak.lat != null ? r(peak.lat, 6) : null,
      lng: peak.lng != null ? r(peak.lng, 6) : null,
      time: peak.time.toISOString().slice(0, 19),
    };
    const tup = peak.jump_tuple;
    if (tup && tup.length >= 4 && tup[0] && tup[0] > 0) {
      jump.airtime_s = r(Number(tup[1]), 2);
      jump.distance_m = r(Number(tup[2]), 1);
      jump.approach_speed_kn = r(Number(tup[3]) * MS_TO_KNOTS, 1);
    } else if (peak.speed_ms) {
      jump.approach_speed_kn = r(peak.speed_ms * MS_TO_KNOTS, 1);
    }
    const chart = peak.jumpchart;
    if (chart && chart.length) {
      const vals = chart.map((v) => v / 100);
      let last = -1;
      vals.forEach((v, i) => { if (v > 0) last = i; });
      if (last >= 0) {
        jump.trajectory_m = [0.0, ...vals.slice(1, last + 1).map((v) => r(v, 2)), 0.0];
      }
    }
    const ts = peak.jump_ts;
    if (ts && ts.length >= 3 && ts[1] && ts[2]) {
      jump.takeoff_ts = Math.trunc(ts[1]);
      jump.landing_ts = Math.trunc(ts[2]);
    }
    jumps.push(jump);
  }
  jumps.sort((a, b) => b.height_m - a.height_m);
  return jumps;
}

/** (unix_ts, lat, lng) for every GPS-fixed record. Port of _read_track. */
function readTrack(records: any[]): Array<[number, number, number]> {
  const track: Array<[number, number, number]> = [];
  for (const rec of records) {
    const lat = semicirclesToDeg(rec.positionLat);
    const lng = semicirclesToDeg(rec.positionLong);
    const t: Date | undefined = rec.timestamp;
    if (lat && lng && t != null) track.push([Math.trunc(t.getTime() / 1000), lat, lng]);
  }
  return track;
}

/** Attach the flight-window GPS path to the top N jumps. Port of attach_jump_paths. */
export function attachJumpPaths(jumps: Jump[], track: Array<[number, number, number]>, topN = 3): void {
  if (!track.length) return;
  for (const j of jumps.slice(0, topN)) {
    const t0 = j.takeoff_ts, t1 = j.landing_ts;
    if (!t0 || !t1) continue;
    const window: Array<[number, number, number]> = [];
    for (const [ts, lat, lng] of track) {
      if (t0 - PATH_LEAD_S <= ts && ts <= t1 + PATH_LEAD_S) window.push([ts - t0, r(lat, 6), r(lng, 6)]);
    }
    if (window.length >= 3) j.path = window;
  }
}

export interface SurfrJump { rank: number; height_m: number; airtime_s: number; distance_m: number; approach_speed_kn: number; }

/** Parse Surfr's 'Top 5 Jumps' block. Port of parse_surfr_description. */
export function parseSurfrDescription(description: string | null | undefined): SurfrJump[] {
  if (!description) return [];
  const out: SurfrJump[] = [];
  for (const m of description.matchAll(SURFR_JUMP_RE)) {
    out.push({
      rank: parseInt(m[1], 10),
      height_m: parseFloat(m[2]),
      airtime_s: parseFloat(m[3]),
      distance_m: parseInt(m[4], 10),
      approach_speed_kn: r(parseInt(m[5], 10) * KMH_TO_KNOTS, 1),
    });
  }
  return out;
}

export interface KitePayload { jumps: Jump[]; summary: Record<string, any>; }

/** Rank + Surfr-enrich jumps and compute the session summary. Port of assemble_jumps. */
export function assembleJumps(jumps: Jump[], surfrDescription: string | null = null): KitePayload {
  const surfr = parseSurfrDescription(surfrDescription);
  jumps.forEach((j, i) => {
    j.rank = i + 1;
    const s = surfr.find((x) => x.rank === i + 1);
    if (s && Math.abs(s.height_m - j.height_m) <= SURFR_MATCH_TOL_M) {
      if (j.airtime_s === undefined) j.airtime_s = s.airtime_s;
      if (j.distance_m === undefined) j.distance_m = s.distance_m;
      j.surfr_height_m = s.height_m;
    }
  });

  let spot: string | null = null;
  let surfrMaxAir: number | null = null;
  if (surfrDescription) {
    const ms = SURFR_SPOT_RE.exec(surfrDescription);
    if (ms) spot = ms[1];
    const ma = SURFR_MAXAIR_RE.exec(surfrDescription);
    if (ma) surfrMaxAir = parseFloat(ma[1]);
  }
  const airtimes = jumps.filter((j) => j.airtime_s).map((j) => j.airtime_s as number);
  const maxAirtime = airtimes.length ? Math.max(...airtimes) : surfrMaxAir;

  const heights = jumps.map((j) => j.height_m);
  const summary = {
    jump_count: jumps.length,
    max_height_m: heights.length ? Math.max(...heights) : null,
    avg_height_m: heights.length ? r(heights.reduce((a, b) => a + b, 0) / heights.length, 2) : null,
    total_height_m: heights.length ? r(heights.reduce((a, b) => a + b, 0), 1) : null,
    max_airtime_s: maxAirtime,
    spot,
    surfr_matched: surfr.length > 0,
  };
  return { jumps, summary };
}

/** Full extraction from FIT bytes + optional Surfr text. Port of build_kite_jumps. */
export function buildKiteJumps(fitBytes: Uint8Array, surfrDescription: string | null = null): KitePayload {
  const { records, devIdx } = decodeRecords(fitBytes);
  const jumps = groupJumps(readJumpSamples(records, devIdx));
  attachJumpPaths(jumps, readTrack(records));
  return assembleJumps(jumps, surfrDescription);
}

// ---- DB + download orchestration (wiring) ----

/**
 * Find the Surfr-exported Strava activity within 15 min of this Garmin start and
 * return its corrected description. Port of _surfr_description. (strava_raw_data
 * is currently stale — this returns null in practice, and the native CIQ jump
 * data is Surfr-accurate anyway.)
 */
export async function surfrDescription(sql: QueryFn, startGmt: string | null): Promise<string | null> {
  if (!startGmt) return null;
  const rows = await sql`
    SELECT raw_json->>'description' AS description
    FROM strava_raw_data
    WHERE jsonb_typeof(raw_json) = 'object'
      AND raw_json->>'name' ILIKE '%surfr%'
      AND raw_json->>'description' ILIKE '%Top%Jump%'
      AND abs(EXTRACT(EPOCH FROM ((raw_json->>'start_date')::timestamptz - (${startGmt} || ' UTC')::timestamptz))) < 900
    ORDER BY abs(EXTRACT(EPOCH FROM ((raw_json->>'start_date')::timestamptz - (${startGmt} || ' UTC')::timestamptz)))
    LIMIT 1`;
  return rows.length ? (rows[0].description ?? null) : null;
}

/** Download the ORIGINAL activity FIT (a zip) and return the .fit bytes. */
export async function downloadFit(client: GarminClient, activityId: number): Promise<Uint8Array> {
  const zipBytes = await client.getBytes(`/download-service/files/activity/${activityId}`);
  const files = unzipSync(zipBytes);
  const fitName = Object.keys(files).find((n) => n.toLowerCase().endsWith(".fit"));
  if (!fitName) throw new Error(`No .fit in download zip for activity ${activityId}`);
  return files[fitName];
}

/**
 * If the activity is a kiteboarding session, extract its per-jump data and store
 * it under endpoint 'kite_jumps'. Port of store_kite_jumps_for_activity +
 * _maybe_extract_kite_jumps. Returns the payload, or null if not a kite activity.
 * DB + a Garmin FIT download. Guarded by the caller so it never breaks ingest.
 */
export async function extractKiteJumpsForActivity(
  sql: QueryFn, client: GarminClient, activityId: number, typeKey: string | null, startGmt: string | null,
): Promise<KitePayload | null> {
  if (!typeKey || !typeKey.toLowerCase().includes("kite")) return null;
  // Dedup: a completed activity's jumps never change, so skip the FIT download
  // if we've already extracted them (mirrors syncActivityDetails' 'details' skip).
  const seen = await sql`
    SELECT 1 FROM garmin_activity_raw WHERE activity_id = ${activityId} AND endpoint_name = 'kite_jumps' LIMIT 1`;
  if (seen.length) return null;
  const fitBytes = await downloadFit(client, activityId);
  const desc = await surfrDescription(sql, startGmt);
  const payload = buildKiteJumps(fitBytes, desc);
  await sql`
    INSERT INTO garmin_activity_raw (activity_id, endpoint_name, raw_json, synced_at)
    VALUES (${activityId}, 'kite_jumps', ${JSON.stringify(payload)}::jsonb, NOW())
    ON CONFLICT (activity_id, endpoint_name)
    DO UPDATE SET raw_json = EXCLUDED.raw_json, synced_at = NOW()`;
  return payload;
}
